apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/DatastoreDbBuilder.cs

```diff
@@ -0,0 +1,247 @@
+                    GaxPreconditions.CheckState(clientBuilder.TokenAccessMethod == null, "{0} should not be set when connecting to an emulator", nameof(clientBuilder.TokenAccessMethod));
+                    clientBuilder.Endpoint = emulatorConfig.Endpoint;
+                    clientBuilder.ChannelCredentials = Grpc.Core.ChannelCredentials.Insecure;
+                    projectId = projectId ?? emulatorConfig.ProjectId;
+                }
+            }
+            GaxPreconditions.CheckState(!string.IsNullOrEmpty(projectId), "The project ID must be configured");
+            return new ConfiguredBuilder(projectId, NamespaceId, clientBuilder);
+        }
+
+        /// <summary>
+        /// The namespace ID, or null to use the default.
+        /// </summary>
+        public string NamespaceId { get; set; }
+
+        /// <inheritdoc />
+        public override DatastoreDb Build() =>
+            PrepareBuilder().Build();
+
+        /// <inheritdoc />
+        public override Task<DatastoreDb> BuildAsync(CancellationToken cancellationToken = default) =>
+            PrepareBuilder().BuildAsync(cancellationToken);
+
+        // We never end up using these methods, at least with the current implementation
+        /// <inheritdoc />
+        protected override ServiceEndpoint GetDefaultEndpoint() =>
+            throw new InvalidOperationException($"This method should never execute in DatastoreDbBuilder");
+
+        /// <inheritdoc />
+        protected override IReadOnlyList<string> GetDefaultScopes() =>
+            throw new InvalidOperationException($"This method should never execute in DatastoreDbBuilder");
+
+        /// <inheritdoc />
+        protected override ChannelPool GetChannelPool() =>
+            throw new InvalidOperationException($"This method should never execute in DatastoreDbBuilder");
+
+        private class EmulatorConfiguration
+        {
+            internal const string EmulatorHostVariable = "DATASTORE_EMULATOR_HOST";
+            internal const string EmulatorProjectVariable = "DATASTORE_PROJECT_ID";
+            internal ServiceEndpoint Endpoint { get; }
+            internal string ProjectId { get; }
+
+            private EmulatorConfiguration(ServiceEndpoint endpoint, string projectId)
+            {
+                Endpoint = endpoint;
+                ProjectId = projectId;
+            }
+
+            internal static EmulatorConfiguration FromEnvironment()
+            {
+                // Treat present-but-empty environment variables as if they were absent.
+                string hostAndPort = Environment.GetEnvironmentVariable(EmulatorHostVariable)?.Trim();
+                string projectId = Environment.GetEnvironmentVariable(EmulatorProjectVariable)?.Trim();
+                if (hostAndPort == "")
+                {
+                    hostAndPort= null;
+                }
+                if (projectId == "")
```

Comment:
nit: And this could be moved to `projectId == "" ? null : projectId` in the ctor call.

Reply:
Done.